src/main/kotlin/com/hj/leetcode/kotlin/problem808/Solution.kt

```diff
@@ -0,0 +1,52 @@
+package com.hj.leetcode.kotlin.problem808
+
+/**
+ * LeetCode page: [808. Soup Servings](https://leetcode.com/problems/soup-servings/);
+ */
+class Solution {
+    /* Complexity:
+     * Time O(1) and Space O(1) since the probability will reach 1.0-tolerance when n is
+     * greater than certain value n0;
+     */
+    fun soupServings(n: Int): Double {
+        val memoization = hashMapOf<RemainingSoup, Double>()
+        val tolerance = 1e-5
+        val unitServeMl = 25
+        for (amount in 0..n step unitServeMl) {
+            if (probabilityAsked(RemainingSoup(amount, amount), memoization) > 1 - tolerance) {
+                return 1.0
+            }
+        }
+        return probabilityAsked(RemainingSoup(n, n), memoization)
+    }
+
+    private fun probabilityAsked(
+        remainingSoup: RemainingSoup,
+        memoization: MutableMap<RemainingSoup, Double> = hashMapOf()
+    ): Double = with(remainingSoup) {
+        if (this in memoization) {
+            return checkNotNull(memoization[this])
+        }
+        if (typeA > 0 && typeB <= 0) {
+            return 0.0
+        }
+        if (typeA <= 0) {
+            return if (typeB <= 0) 0.5 else 1.0
+        }
+
+        val result = 0.25 * (probabilityAsked(remainingSoup.afterOperation1(), memoization)
+                + probabilityAsked(remainingSoup.afterOperation2(), memoization)
+                + probabilityAsked(remainingSoup.afterOperation3(), memoization)
+                + probabilityAsked(remainingSoup.afterOperation4(), memoization))
+        memoization[remainingSoup] = result
+        return result
+    }
+
+    private data class RemainingSoup(val typeA: Int, val typeB: Int) {
```

Comment:
convey unit of remaining soup, for example, rename to `RemainingSoupMl`

Reply:
Resolved at https://github.com/hj-core/hj-leetcode-kotlin/pull/114/commits/3d034a43c45eb67267411b234297881a11624785